### PR TITLE
Stop pipeline

### PR DESCRIPTION
**Pull request by abharathkr**
# Problem
referenced feature(s): #643 

# Solution
Provided a "Stop Pipeline" button in consortia page for active pipeline
# Testing

### CODE REVIEW

File: packages/coinstac-client-core/src/index.js [hunk 1]
```diff
@@ -237,6 +237,19 @@ class CoinstacClient {
...
 
+  /**
+   * Stop pipeline
+   * @param {string} pipelineId The id of the pipeline running
+   * @param {string} runId The id of the pipeline run
+   */
+
+  stopPipeline( pipelineId, runId ) {
+    const stopPipeline = this.pipelineManager.stopPipeline(pipelineId, runId);
+    return new Promise((res, rej) => {
```
Comment (rssk) [inline packages/coinstac-client-core/src/index.js]: Why create a new promise here? Also, this new promise doesnt resolve or reject anything, so you'll never actually get the async data to return right. `stopPipeline` should return a promise and thus you can just make this a one liner es6 arrow function

File: packages/coinstac-pipeline/src/controller.js [hunk 2]
```diff
@@ -235,6 +239,10 @@ module.exports = {
           trampoline(() => {
             return queue.length
               ? function _cb(...args) {
+                console.log(controllerState);
```
Comment (rssk) [inline packages/coinstac-pipeline/src/controller.js]: ;-)

File: packages/coinstac-pipeline/src/controller.js [hunk 3]
```diff
@@ -235,6 +239,10 @@ module.exports = {
...
+                if (controllerState.stopByUser == 'stop'){
+                  err({message: 'Pipeline run is stopped by user'});
```
Comment (rssk) [inline packages/coinstac-pipeline/src/controller.js]: It's good practice to actually create an error when error'ing out, as often upstream code checks for object type. 
`err(new Error('Pipeline run is stopped by user'))`

File: packages/coinstac-pipeline/src/pipeline-manager.js [hunk 4]
```diff
@@ -541,6 +541,11 @@ module.exports = {
 
         return this.activePipelines[runId].stateEmitter;
       },
+      stopPipeline(pipelineId, runId) {
+        this.activePipelines[runId].pipeline.pipelineSteps.forEach( step => {
```
Comment (rssk) [inline packages/coinstac-pipeline/src/pipeline-manager.js]: Couple things here, the pipeline-manager here is directly mussing w/ pipeline objects, I try to do this only when absolutely necessary. Also, there is only ever one currently running pipeline `step` and so you would only need to ever stop that one. 
What needs to happen is that the call to stop the currently running step needs to happen from manager to a pipeline function, and that function needs to call the currently active step's controller, thus stopping the pipeline.

Comment (rssk) [reply packages/coinstac-pipeline/src/pipeline-manager.js]: Also also, docbloc your functions :-)

File: packages/coinstac-ui/app/main/index.js [hunk 5]
```diff
@@ -238,6 +238,19 @@ loadConfig()
...
+    /**
+   * IPC Listener to stop pipeline
+   * @param {String} pipelineId The id of the pipeline currently running
+   * @param {String} runId The id of the pipeline run
+   * @return {Promise<String>} Status message
+   */
+
+    ipcMain.on('stop-pipeline', (event, { pipelineId, runId }) => {
+      return core.stopPipeline(pipelineId, runId)
+      .catch((err) => { console.log(`Pipeline could not be stopped because of the following error: ${err}`)});
```
Comment (rssk) [inline packages/coinstac-ui/app/main/index.js]: this should probably be logged using the winston instance and then sent to the UI toast handler

**Review by rssk — CHANGES_REQUESTED** [review]

**Comment by rssk** [comment]
Thanks for cleaning this up, one more quick task:

Can you add a stop button to the home page where the pipeline running widget is?

**Comment by rochaeb** [comment]
> Thanks for cleaning this up, one more quick task:
> 
> Can you add a stop button to the home page where the pipeline running widget is?

done

**Pull request merged**